leptonai/util/cancel_on_disconnect.py

```diff
@@ -0,0 +1,101 @@
+"""
+Implements the functionality to cancel a fastapi call when the client disconnects.
+"""
+
+import anyio
+from exceptiongroup import catch, ExceptionGroup
+from fastapi import Request, HTTPException
+from loguru import logger
+
+
+class ClientDisconnected(Exception):
+    """
+    Exception raised when the client disconnects.
+    """
+
+    pass
+
+
+class TaskFinished(Exception):
+    """
+    Exception raised when the task finishes.
+    """
+
+    pass
+
+
+async def _disconnect_poller(request: Request, interval: float):
+    """
+    Poll for disconnect. If the request disconnects, stop polling and return.
+    """
+    logger.trace("Starting disconnect poller")
+    try:
+        while True:
+            await anyio.sleep(interval)
+            if await request.is_disconnected():
+                logger.trace(f"Request {request} disconnected.")
+                raise ClientDisconnected()
+    except anyio.get_cancelled_exc_class():
+        logger.trace("Stopping disconnect poller")
+
+
+async def run_with_cancel_on_disconnect(
+    request: Request, cancel_on_connect_interval: float, callback, *args, **kwargs
+):
+    """
+    Run the callback with the given arguments. If the client disconnects, cancel the callback.
+
+    Args:
+        request: The request object.
+        cancel_on_connect_interval: The interval in seconds to poll for disconnect.
+        callback: The callback to run.
+        args: The arguments to pass to the callback.
+        kwargs: The keyword arguments to pass to the callback.
+    """
+    result = []
+
+    async def _callback_task_wrapper():
+        logger.trace("Starting callback task wrapper")
+        try:
+            ret = await callback(*args, **kwargs)
+            result.append(ret)
+            raise TaskFinished()
+        except anyio.get_cancelled_exc_class():
+            # If the task is cancelled, do not raise anything.
+            logger.trace("Callback task cancelled.")
+        except TaskFinished:
+            # If the task is finished, simply raise to finish execution.
+            logger.trace("Callback task finished.")
+            raise
+        except Exception as e:
+            # For any user code exceptions, we want to raise them and return.
+            logger.trace(f"Callback task exception: {e}")
+            raise HTTPException(status_code=500, detail=str(e))
+
+    def handle_task_finished(e: ExceptionGroup) -> None:
```

Comment:
IIUC the "normal" execution path is raising `TaskFinished` at line 62, and then goes to line 66, log a line at 68, reraise again at line 69, goes to line 92, and then here inside handle_task_finished, logs a line and check `result` is not empty.
Can we not raise TaskFinished at all, and then checks the `result` list is not empty at line 99 (i.e. after the `with catch` block)?

Reply:
Actually we do need to raise `TaskFinished` somewhere - that is to tell the taskgroup "hey, I finished, let's cancel that disconnect poller". Without it, the taskgroup will continuously wait for the disconnect poller and never return.

This uses the taskgroup's feature that if one of the tasks throws an exception, the group automatically cancels all other tasks.

But you did find an issue - I don't really need to raise `TaskFinished` at line 62. I should put it outside the try-catch pair so I don't need to do the raise-catch-reraise thingy. Let me fix it.

Reply:
I see. I was thinking we can explicitly call `.cancel` on the disconnect poller task. But yeah this should work.